##  1. Importing important libraries

In [19]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

##  2. Importing data

In [2]:
data1 = pd.read_csv('bank-full.csv',sep=';')
data1.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no

## 3. Data understanding

In [3]:
data1.dtypes

age           int64
job          object
marital      object
education    object
default      object
balance       int64
housing      object
loan         object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
y            object
dtype: object

In [4]:
data1.shape

(45211, 17)

In [5]:
data1.isnull().sum()

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

## 4. Data Preparation

In [6]:
le = LabelEncoder()
data1['y'] = le.fit_transform(data1['y'])

In [7]:
data1['job'] = le.fit_transform(data1['job'])
data1['marital'] = le.fit_transform(data1['marital'])
data1['education'] = le.fit_transform(data1['education'])
data1['default'] = le.fit_transform(data1['default'])
data1['housing'] = le.fit_transform(data1['housing'])
data1['contact'] = le.fit_transform(data1['contact'])
data1['month'] = le.fit_transform(data1['month'])
data1['poutcome'] = le.fit_transform(data1['poutcome'])
data1['loan'] = le.fit_transform(data1['loan'])

In [8]:
data1.head()

age  job  marital  education  default  balance  housing  loan  contact  \
0   58    4        1          2        0     2143        1     0        2   
1   44    9        2          1        0       29        1     0        2   
2   33    2        1          1        0        2        1     1        2   
3   47    1        1          3        0     1506        1     0        2   
4   33   11        2          3        0        1        0     0        2   

   day  month  duration  campaign  pdays  previous  poutcome  y  
0    5      8       261         1     -1         0         3  0  
1    5      8       151         1     -1         0         3  0  
2    5      8        76         1     -1         0         3  0  
3    5      8        92         1     -1         0         3  0  
4    5      8       198         1     -1         0         3  0

In [9]:
data1.dtypes

age          int64
job          int32
marital      int32
education    int32
default      int32
balance      int64
housing      int32
loan         int32
contact      int32
day          int64
month        int32
duration     int64
campaign     int64
pdays        int64
previous     int64
poutcome     int32
y            int32
dtype: object

## 5. Model Building

In [10]:
X = data1.drop('y',axis=1)
y = data1[['y']]

In [11]:
x_train,x_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state=12)

In [12]:
print(x_train.shape)
print(y_train.shape)

(36168, 16)
(36168, 1)


In [13]:
print(x_test.shape)
print(y_test.shape)

(9043, 16)
(9043, 1)


## 6. Model Training 

In [14]:
logistic_model = LogisticRegression()
logistic_model.fit(x_train,y_train)

C:\Users\namrata\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
C:\Users\namrata\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

## 7. Model Testing 

In [15]:
# checking training accuracy 
y_pred_train = logistic_model.predict(x_train)
y_pred_train

array([0, 0, 1, ..., 0, 0, 1])

In [16]:
## Checking training accuracy
y_pred_test = logistic_model.predict(x_test)
y_pred_test

array([0, 0, 0, ..., 0, 0, 0])

## 8.Model Evaluation

In [17]:
## training accuracy
accuracy_score(y_train,y_pred_train)

0.8866677726166777

In [18]:
## testing accuracy 
accuracy_score(y_test,y_pred_test)

0.8895278115669578

In [20]:
confusion_matrix(y_train,y_pred_train)

array([[31355,   574],
       [ 3525,   714]], dtype=int64)

In [21]:
confusion_matrix(y_test,y_pred_test)

array([[7851,  142],
       [ 857,  193]], dtype=int64)

###### From above, it is clear that total 7851+193 correct predictions and  857+142 are incorrect predictions are done by the model 

In [22]:
print(classification_report(y_train,y_pred_train))

              precision    recall  f1-score   support

           0       0.90      0.98      0.94     31929
           1       0.55      0.17      0.26      4239

    accuracy                           0.89     36168
   macro avg       0.73      0.58      0.60     36168
weighted avg       0.86      0.89      0.86     36168

